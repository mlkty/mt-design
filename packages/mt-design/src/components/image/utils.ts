
export const isValidImage = (src: string) => {
    return new Promise(resolve => {
        const image = new Image();
        image.onload = () => resolve(true);
        image.onerror = () => resolve(false);
        image.src = src;
    });
};
