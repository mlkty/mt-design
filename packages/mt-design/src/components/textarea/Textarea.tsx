import {c, useControlled, useLayoutEffect} from '@mlkty/mt-shared-utils';
import {CSSProperties, ForwardedRef, forwardRef, useImperativeHandle, useRef, useState} from 'react';
import {useConfigContext} from '../config-provider';

export type TextareaProps = React.DetailedHTMLProps<
    React.TextareaHTMLAttributes<HTMLTextAreaElement>,
    HTMLTextAreaElement
> & {
    rows?: number;
    /**
     * 响应式撑高。默认高度为 rows。
     */
    autoSize?: boolean | {minRows?: number, maxRows?: number};
    disabled?: boolean;
    showCount?: boolean;
    value?: string;
    defaultValue?: string;
    placeholder?: string;
    maxLength?: number;
    onChange?: (val: string, prev: string) => void;

    rootClassName?: string;
    rootStyle?: CSSProperties;
    className?: string;
    style?: CSSProperties;
};

export interface TextareaRef {
    clear: () => void;
    focus: () => void;
    blur: () => void;
    nativeElement: HTMLTextAreaElement | null;
}

const Textarea = forwardRef((props: TextareaProps, ref: ForwardedRef<TextareaRef>) => {
    const {
        rows = 2,
        placeholder,
        showCount,
        maxLength,
        autoSize,
        disabled,

        // events
        onChange,
        onFocus,
        onBlur,
        onCompositionStart,
        onCompositionEnd,

        // value
        defaultValue: outerDefaultValue,
        value: outerValue,

        // style
        rootClassName,
        rootStyle: outerRootStyle,
        className,
        style: outerStyle,

        ...restProps
    } = props;

    const {getPrefixCls} = useConfigContext();
    const nativeTextAreaRef = useRef<HTMLTextAreaElement>(null);
    const hiddenTextAreaRef = useRef<HTMLTextAreaElement>(null);
    const compositingRef = useRef(false);
    const heightRef = useRef<string>('auto');
    const [height, setHeight] = useState<string>('auto');
    const [isFocus, setIsFocus] = useState(false);
    const [isHover, setIsHover] = useState(false);

    const [value, setValue] = useControlled('', {
        defaultValue: outerDefaultValue,
        value: outerValue,
        onChange,
    });

    useImperativeHandle(ref, () => ({
        clear: () => {
            setValue('');
        },
        focus: () => {
            nativeTextAreaRef.current?.focus();
        },
        blur: () => {
            nativeTextAreaRef.current?.blur();
        },
        get nativeElement() {
            return nativeTextAreaRef.current;
        },
    }));

    useLayoutEffect(() => {
        if (!autoSize) {
            return;
        }
        const textArea = nativeTextAreaRef.current;
        const hiddenTextArea = hiddenTextAreaRef.current;
        if (!textArea || !hiddenTextArea) {
            setHeight('auto');
            return;
        }
        let height = hiddenTextArea.scrollHeight;
        if (typeof autoSize === 'object') {
            const computedStyle = window.getComputedStyle(textArea);
            const lineHeight = parseFloat(computedStyle.lineHeight);
            if (autoSize.minRows) {
                height = Math.max(height, autoSize.minRows * lineHeight);
            }
            if (autoSize.maxRows) {
                height = Math.min(height, autoSize.maxRows * lineHeight);
            }
        }
        // @NOTE 需要继续探究一下为什么存在 2px 误差。
        heightRef.current = `${height}px`;
        setHeight(heightRef.current);
    }, [value, autoSize]);

    const prefixCls = getPrefixCls('textarea');


    const rootCls = c(prefixCls, {
        [`${prefixCls}--focus`]: isFocus,
        [`${prefixCls}--hover`]: isHover,
        [`${prefixCls}--disabled`]: disabled,
    }, rootClassName);

    const rootStyle = {
        ...outerRootStyle,
    };

    const cls = c(`${prefixCls}-element`, className);

    const style = {
        height,
        ...outerStyle,
    };

    const renderCount = () => {
        if (!showCount) {
            return null;
        }

        let text = '';

        if (showCount) {
            text += value.length;
        }

        if (maxLength) {
            text += ` / ${maxLength}`;
        }
        return <div className={`${prefixCls}-count`}>{text}</div>;
    };

    return (
        <div className={rootCls} style={rootStyle}>
            <textarea
                {...restProps}
                key="text-area"
                ref={nativeTextAreaRef}
                className={cls}
                style={style}
                value={value}
                onChange={e => {
                    let v = e.target.value;
                    if (maxLength && !compositingRef.current) {
                        v = [...v].slice(0, maxLength).join('');
                    }
                    setValue(v);
                }}
                onFocus={e => {
                    setIsFocus(true);
                    props.onFocus?.(e);
                }}
                onBlur={e => {
                    setIsFocus(false);
                    props.onBlur?.(e);
                }}
                onMouseEnter={e => {
                    setIsHover(true);
                    props.onMouseEnter?.(e);
                }}
                onMouseLeave={e => {
                    setIsHover(false);
                    props.onMouseLeave?.(e);
                }}
                onCompositionStart={e => {
                    compositingRef.current = true;
                    props.onCompositionStart?.(e);
                }}
                onCompositionEnd={e => {
                    compositingRef.current = false;
                    if (maxLength) {
                        const v = (e.target as HTMLTextAreaElement).value;
                        setValue([...v].slice(0, maxLength).join(''));
                    }
                    props.onCompositionEnd?.(e);
                }}
                autoComplete={props.autoComplete}
                autoFocus={props.autoFocus}
                disabled={props.disabled}
                readOnly={props.readOnly}
                name={props.name}
                onClick={props.onClick}
                placeholder={placeholder}
                id={props.id}
                rows={rows}
            />
            {renderCount()}
            {autoSize && (
                <textarea
                    ref={hiddenTextAreaRef}
                    className={`${prefixCls}-element ${prefixCls}-element--hidden`}
                    value={value}
                    rows={props.rows}
                    aria-hidden
                    readOnly
                />
            )}
        </div>
    );
});

export default Textarea;
