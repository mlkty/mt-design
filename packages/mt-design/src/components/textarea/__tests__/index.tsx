import {createRef} from 'react';
import {Textarea} from '..';
import {act, customRender, waitFor, fireEvent} from '../../../../tests/utils';
import mountTest from '../../../../tests/mountTest';
import {TextareaRef} from '../Textarea';

describe('Textarea', () => {
    mountTest(Textarea);

    it('render single textarea', async () => {
        const {container} = customRender(<Textarea />);
        expect(container).toMatchSnapshot();
    });

    it('render autoSize textarea', async () => {
        const {container} = customRender(<Textarea autoSize />);
        expect(container).toMatchSnapshot();
    });

    it('render rows property', async () => {
        const {container} = customRender(<Textarea rows={5} />);
        expect(container.querySelector('textarea')).toHaveAttribute('rows', '5');
    });

    it('render disabled property', async () => {
        const {container} = customRender(<Textarea disabled />);
        expect(container.querySelector('textarea')).toHaveAttribute('disabled');
        expect(container.firstChild).toHaveClass('mt-textarea--disabled');
    });

    it('render showCount property', async () => {
        const {container} = customRender(<Textarea showCount defaultValue="12345" />);
        expect(container.querySelector('.mt-textarea-count')).toHaveTextContent('5');
    });

    it('render hover status', async () => {
        const {container} = customRender(<Textarea />);
        fireEvent.mouseEnter(container.querySelector('textarea')!);
        expect(container.firstChild).toHaveClass('mt-textarea--hover');
        fireEvent.mouseLeave(container.querySelector('textarea')!);
        expect(container.firstChild).not.toHaveClass('mt-textarea--hover');
    });

    it('render focus status', async () => {
        const {container} = customRender(<Textarea />);
        fireEvent.focus(container.querySelector('textarea')!);
        expect(container.firstChild).toHaveClass('mt-textarea--focus');
        fireEvent.blur(container.querySelector('textarea')!);
        expect(container.firstChild).not.toHaveClass('mt-textarea--focus');
    });

    it('render composition status', async () => {
        const text = '123456789101112';
        const {container} = customRender(<Textarea maxLength={10} />);
        fireEvent.compositionStart(container.querySelector('textarea')!);
        fireEvent.change(container.querySelector('textarea')!, {
            target: {value: text},
        });
        expect(container.querySelector('textarea')).toHaveValue(text);
        fireEvent.compositionEnd(container.querySelector('textarea')!);
        expect(container.querySelector('textarea')).toHaveValue(text.slice(0, 10));
    });

    it('render maxLength', async () => {
        const text = '123456789101112';
        const {container} = customRender(<Textarea maxLength={10} />);
        fireEvent.change(container.querySelector('textarea')!, {
            target: {value: text},
        });
        expect(container.querySelector('textarea')).toHaveValue(text.slice(0, 10));
    });

    it('render showCount & maxLength property', async () => {
        const {container} = customRender(<Textarea showCount maxLength={100} defaultValue="12345" />);
        expect(container.querySelector('.mt-textarea-count')).toHaveTextContent('5 / 100');
    });

    it('using ref', async () => {
        const ref = createRef<TextareaRef>();
        const {container} = customRender(<Textarea ref={ref} defaultValue="12345" />);

        act(() => ref.current?.focus());
        waitFor(() => expect(container.firstChild).toHaveClass('mt-textarea--focus'));

        act(() => ref.current?.blur());
        waitFor(() => expect(container.firstChild).not.toHaveClass('mt-textarea--focus'));

        act(() => ref.current?.clear());
        waitFor(() => expect(ref.current?.nativeElement).toHaveTextContent(''));
    });

    it('render autoSize property', async () => {
        const {container} = customRender(<Textarea defaultValue={'1\n2\n3\n4\n5\n6'} style={{lineHeight: 24}} autoSize={{minRows: 2, maxRows: 3}} />);
        // @TODO 理应 72px
        expect(container.querySelector('textarea')).toHaveStyle('height: 48px');
    });

    it('render onChange events', async () => {
        const onChange = jest.fn();
        const {container} = customRender(<Textarea value="1" onChange={onChange} />);
        fireEvent.change(container.querySelector('textarea')!, {
            target: {value: 'abc'},
        });
        expect(onChange).toBeCalledTimes(1);
        expect(onChange).toBeCalledWith('abc', '1');
    });
});
