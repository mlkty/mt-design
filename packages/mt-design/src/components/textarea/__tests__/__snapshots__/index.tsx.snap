// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Textarea render autoSize textarea 1`] = `
<div>
  <div
    class="mt-textarea"
  >
    <textarea
      class="mt-textarea-element"
      rows="2"
      style="height: 0px;"
    />
    <textarea
      aria-hidden="true"
      class="mt-textarea-element mt-textarea-element--hidden"
      readonly=""
    />
  </div>
</div>
`;

exports[`Textarea render single textarea 1`] = `
<div>
  <div
    class="mt-textarea"
  >
    <textarea
      class="mt-textarea-element"
      rows="2"
      style="height: auto;"
    />
  </div>
</div>
`;
